docs/developer-docs/security/general-security-best-practices.md

```diff
@@ -2,42 +2,106 @@
+
+## General Best Practices
+
+### 1. Generous canister top-up
+
+Ensure all canisters have sufficient cycles to sustain operations for several years initially. Given the cost-effectiveness of storage and computation on ICP compared to other platforms, this upfront investment is usually modest.
+
+Refer to [topping up cycles page](../production/topping-up-canister.md) to learn how to top-up cycles effectively.
+
+Refer to the [cycles estimate](../gas-cost.md) page to estimate cycle usage. 
+
+### 2. Managing cycle depletion
+
+ICP features a mechanism to prevent canisters from running out of cycles. Canisters have a configurable `freezing_threshold`, dynamically evaluated in cycles. Set `freezing_threshold` conservatively, ensuring at least 90 to 180 days' worth of cycles for proactive management.
+
+Refer to [this resource](../production/topping-up-canister.md#managing-cycle-depletion-with-freezing-threshold) to learn how to configure the `freezing_threshold` of a canister. 
+
+To make sure you won’t get surprised by a high cycle burn rate or hitting an instruction limit, you can use the recently added [performance counter API](https://internetcomputer.org/blog/features/async-performance-counter) to profile your canisters even before going live.
+
+### 3. Efficient Implementations
+
+Optimize canister implementation to avoid unnecessary costs. 
+
+Use [canister timers](../backend/periodic-tasks.md) over plain heartbeats for reduced daily costs.
 
+#### Motoko
 
-## Certify query responses if they are relevant for security
+1. Opt for `TrieMap` over `HashMap` to prevent automatic resizing overhead.
+2. Utilize `Buffer` instead of `Array` for dynamic resizing.
+3. Consider `Blob` over `[Nat8]` for compact storage and reduced GC pressure:
 
-### Security concern
+ * Use `Blob` instead of `[Nat8]` for storing large binary assets.
+ * Use `Blob` instead of `[Nat8]` when sending or receiving Candid `vec nat8/blob` values. The choice is yours but `Blobs` are 4x more compact and much less taxing on garbage collection (GC).
+
+4. Store large Blobs in stable memory for efficient manual management.
+5. Use the `compacting-gc` setting, especially in append-only scenarios, to allow access to larger heaps and reduce the cost of copying large, stationary objects.
+
+#### Rust 
+1. Exercise caution with `Vec<u8>` and `String types` for state serialization.
```

Comment:
Could this be made more concrete? What is the risk of using these types, and what exactly should I do as a developer to avoid these issues?